### ISSUE TITLE
support eager loading

### ISSUE THREAD

**Issue by chirag04** [pldin601/Fluorite.js#1]
awesome lib. very simple yet powerful. I think eager loading is the only thing missing for my usecase.

```
const thing = await Thing.objects.get(3);
const user = await thing.user().one();
expect(user.id).toBe(2);
```

It would be nice to get the user for a given `Thing` in a single query via eager loading. I'm not sure what the api might look like for this.

**Comment by pldin601** [comment]
Thanks! 

Maybe using `with`?

```javascript
const thing = await Thing.objects.with('user').get(3);
const user = thing.related('user');
expect(user.id).toBe(2);
```

I'm planning to add eager loading that will use second query to load related object. It is simpler to implement than joining tables.

**Comment by chirag04** [comment]
I guess using second query defeats the purpose of eager loading. It would be awesome to do it with joins. I understand it can get tricky. Idk

**Comment by pldin601** [comment]
I will think about that.

It depends on relation type. For example, if I want to fetch `User` with all of it's `Things` will be effective to execute two SQL statements: fetch user and fetch user's things. But if I want to fetch thing with related user, join is more effective.

I'm working now on relation abstractions and some API changes.
For example it will be possible to iterate over models asynchronously:

```javascript
for await (const user of User.objects) {
  console.log(user.get('name'));
}
```

Also removed `.one` and `.all` methods from query. To evaluate query just use `await` or `then` instead:

```javascript
const allUsers = User.objects;
const adults = User.objects.filter({ age__gte: 18 });

console.log(await allUsers); // will execute statement and return all users
console.log(await adults); // will execute statement and return list of adult users

// Or use as Promise:
User.objects.then(users => console.log(users));
```

After it I planned to implement eager.

**Comment by pldin601** [comment]
It is not possible to implement `belongsTo` with joins using current API due to SQL limitations. 

The `knex` query returns row data with column names but without table names. We need to rename columns adding table information to extract this information from row data and fulfill eager loaded models.

To achieve this each model must contain list of its table column names.

**Comment by chirag04** [comment]
> To achieve this each model must contain list of its table column names.

This is how most ORMs do it. i think it's fair to do it this way. Infact it's better because you don't have to do `select *` and can use the column names to build the select clause.

**Comment by pldin601** [comment]
You right.

Would you like to contribute?

**Comment by chirag04** [comment]
I might not be able to get to it any time soon. had to roll with vanilla `knex` queries for the time being.

**Comment by pldin601** [comment]
Today was released `1.0.1` version. Now supports eager loading relations. At this point it is based on second query principle. Joining will be in a future versions.
